### PR TITLE
Dashboard Header FC, using entity, select, and dispatch hooks

### LINKED ISSUE #39260: Convert Dashboard header to functional component

### PR DESCRIPTION
**Pull request by npfitz**
Closes https://github.com/metabase/metabase/issues/39260

### Description
Migrating the Dashboard Header to be a functional component, leveraging entity hooks, as well as removing Redux Connect. This is largely done as we may have a need to use hooks in this component as part of https://github.com/metabase/metabase/issues/39259 . The experience should not change for the end user (save for a loading state while fetching the collection, which may be _slightly_ different).

### Checklist
Existing tests should cover this component and the functionality inside it

### CODE REVIEW

File: frontend/src/metabase/dashboard/components/DashboardHeader/tests/setup.tsx [hunk 1]
```diff
@@ -121,17 +127,20 @@ export const setup = async ({
...
     toggleSidebar: jest.fn(),
     sidebar: {
-      name: "",
+      name: "" as DashboardSidebarName,
       props: {},
     },
     location: {
       query: {},
-    },
+    } as Location,
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/DashboardHeader/tests/setup.tsx:135]: [`createMockLocation` ? ](https://github.com/metabase/metabase/blob/1f83821e2a2a38a18e24fb938f16dce711b7aa1e/frontend/src/metabase-types/store/mocks/routing.ts#L15C14-L15C32)

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx [hunk 2]
```diff
@@ -5,14 +5,19 @@ import type { Route } from "react-router";
...
+  FetchDashboardResult,
+  SuccessfulFetchDashboardResult,
+} from "metabase/dashboard/types";
 import { isSmallScreen, getMainElement } from "metabase/lib/dom";
 import { FilterApplyButton } from "metabase/parameters/components/FilterApplyButton";
 import SyncedParametersList from "metabase/parameters/components/SyncedParametersList/SyncedParametersList";
 import { getVisibleParameters } from "metabase/parameters/utils/ui";
 import type { EmbeddingParameterVisibility } from "metabase/public/lib/types";
-import type Database from "metabase-lib/metadata/Database";
+//import type Database from "metabase-lib/metadata/Database";
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:20]: Think we can remove this line?

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx [hunk 3]
```diff
@@ -130,10 +131,13 @@ interface DashboardProps {
...
     tabId: DashboardTabId | null;
   }) => void;
+  addHeadingDashCardToDashboard: (opts: NewDashCardOpts) => void;
+  addMarkdownDashCardToDashboard: (opts: NewDashCardOpts) => void;
+  addLinkDashCardToDashboard: (opts: NewDashCardOpts) => void;
   archiveDashboard: (id: DashboardId) => Promise<void>;
 
   onRefreshPeriodChange: (period: number | null) => void;
-  setEditingDashboard: (dashboard: IDashboard) => void;
+  setEditingDashboard: (dashboard: IDashboard | boolean) => void;
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:140]: I'm not sure it should accept a `boolean` at this point :(

File: frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx [hunk 4]
```diff
@@ -630,68 +609,62 @@ class DashboardHeaderContainer extends Component<DashboardHeaderProps> {
...
 
-  saveAsPDF = async () => {
-    const { dashboard } = this.props;
-    const cardNodeSelector = `#${DASHBOARD_PDF_EXPORT_ROOT_ID}`;
+  const saveAsPDF = async () => {
+    const cardNodeSelector = "#Dashboard-Cards-Container";
     await saveDashboardPdf(cardNodeSelector, dashboard.name).then(() => {
       trackExportDashboardToPDF(dashboard.id);
     });
   };
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx:620]: nit: think it'd be way more discoverable if we keep all callbacks next to each other

File: frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx [hunk 5]
```diff
@@ -500,6 +478,7 @@ class DashboardHeaderContainer extends Component<DashboardHeaderProps> {
         />,
       );
 
+      // I don't think this is used anymore
```
Comment (kulyk) [inline frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx:481]: yeah, i think it's only accessible from the "last edit by ..." thing or the "info" sidebar. Let's remove it

**Review by kulyk — APPROVED** [review]
THANK YOU

File: frontend/src/metabase/dashboard/components/DashboardHeader/tests/setup.tsx (hunk 1 above)
Comment (npfitz) [reply frontend/src/metabase/dashboard/components/DashboardHeader/tests/setup.tsx:135]: Ouuuu 🔥

File: frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx (hunk 3 above)
Comment (npfitz) [reply frontend/src/metabase/dashboard/components/Dashboard/Dashboard.tsx:140]: I also thought it was super weird, but when it's passed to the header, we pass "false" when we're done editing :/
https://github.com/metabase/metabase/blob/01fc5ae5a26799d28a5affbceab14bd44b20e1bb/frontend/src/metabase/dashboard/components/DashboardHeader/DashboardHeader.tsx#L254

**Pull request merged**